core/modules/qdisp/testQDisp.cc

```diff
@@ -0,0 +1,412 @@
+    FinishTest() : _finishCalled(false) {}
+    virtual ~FinishTest() {}
+    virtual void operator()(bool val) {
+        _finishCalled = true;
+        LOGF_INFO("_finishCalled=%1%" % _finishCalled);
+    }
+    bool _finishCalled;
+};
+
+class CancelTest : public qdisp::ResponseRequester::CancelFunc {
+public:
+	CancelTest() : _count(0) {}
+	virtual void operator()() {
+		++_count;
+	}
+	int _count;
+};
+
+class ResponseRequesterTest : public qdisp::ResponseRequester {
+public:
+	ResponseRequesterTest() : _code(0), _finished(false) {}
+	virtual std::vector<char>& nextBuffer() {
+		return _vect;
+	}
+	virtual bool flush(int bLen, bool last) {
+		return bLen == magic();
+	}
+	virtual void errorFlush(std::string const& msg, int code) {
+		_msg = msg;
+		_code = code;
+	}
+	virtual bool finished() const {
+		return _finished;
+	}
+	virtual bool reset() {
+		return true;
+	}
+	virtual std::ostream& print(std::ostream& os) const {
+		return os;
+	}
+	static int magic() {return 8;}
+	std::vector<char> _vect;
+	std::string _msg;
+	int _code;
+	bool _finished;
+};
+
+// Start adds 'copies' number of test requests that each sleep for 'millisecs' time
+// before signaling to 'ex' that they are done.
+// Returns time to complete in seconds.
+double executiveTest(qdisp::Executive &ex, SequentialInt &sequence, SequentialInt &chunkId, std::string &millisecs, int copies) {
+	// Test class Executive::add
+	// Modeled after ccontrol::UserQuery::submit()
+	ResourceUnit ru;
+	std::string chunkResultName = "mock";
+	boost::shared_ptr<rproc::InfileMerger> infileMerger;
+	boost::shared_ptr<ChunkMsgReceiverMock> cmr = ChunkMsgReceiverMock::newInstance(chunkId.get());
+	ccontrol::MergingRequester::Ptr mr = boost::make_shared<ccontrol::MergingRequester>(cmr, infileMerger, chunkResultName);
+	std::string msg = millisecs;
+	boost::chrono::system_clock::time_point start = boost::chrono::system_clock::now();
+	std::vector<qdisp::Executive::Spec> s(copies);
+	for (int j=0; j < copies; ++j) {
+		s[j].resource = ru;
+		s[j].request = msg;
+		s[j].requester = mr;
+		ex.add(sequence.incr(), s[j]);
+	}
+	ex.join();
+	boost::chrono::duration<double> secs = boost::chrono::system_clock::now() - start;
+	double ret = secs.count();
+	return ret;
+}
+
+class WorkerTest {
+public:
+	double _retVal;
+	void operator()(qdisp::Executive &ex, SequentialInt &sequence, SequentialInt &chunkId, std::string &millisecs, int copies) {
+		_retVal = executiveTest(ex, sequence, chunkId, millisecs, copies);
+	}
+};
+
+// This function is run in a separate thread to fail the test if it takes too long
+// for the jobs to complete.
+void timeoutFunc(qdisp::Executive &ex, int jobs, int millisecs) {
+	LOGF_INFO("timeoutFunc");
+	boost::this_thread::sleep(boost::posix_time::milliseconds(millisecs));
+	LOGF_INFO("timeoutFunc sleep over %1% empty=%2%" % millisecs % ex._empty.get());
+	BOOST_REQUIRE(ex._empty.get() == true);
+}
+
+BOOST_AUTO_TEST_CASE(Executive) {
```

Comment:
Overall remark: Good job, these mt tests seems complex to code. I didn't understand the very details of the logic but I hope I'll get in that interesting code in the future ;-)